cloud: add isInTrial helper function

TattdCodeMonkey opened the pull request:
## Summary

Added the isInTrial helper function to the cloud plugin setup & start interfaces that allow checking if the deployment or serverless project are in trial based on either the trial_end_date or serverless.in_trial config values being set. Adding a function allows the date comparison to be handled by the cloud plugin and possibly updated in the future without consumers needing to know the details of how we verify if the cluster is within a trial cloud organization.

This is a follow-up to [#232703](https://github.com/elastic/kibana/pull/232703) where I added the `serverless.in_trial` config value. In subsequent PRs it was discussed we should have a helper on the cloud function to give uniform way to access "in trial".

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] ~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)~
- [ ] ~[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials~
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] ~If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)~
- [ ] ~This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.~
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] ~The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)~
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

TattdCodeMonkey committed:
cloud: add getInTrial helper function

Added the getInTrial helper function to the cloud plugin setup & start
interfaces that allow checking if the deployment or serverless project
are in trial based on either the trial_end_date or serverless.in_trial
config values being set. Adding a function allows the date comparison to
be handled by the cloud plugin and possibly updated in the future
without consumers needing to know the details of how we verify if the
cluster is within a trial cloud organization.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/351455)
* Commit: a5bfe92df00802b948318c19e682393530717f9d

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/351455#0199e8ee-45a4-4fec-8cc5-ba22cece5f00)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351455","number":351455,"commit":"a5bfe92df00802b948318c19e682393530717f9d"}],"number":351455}
buildkite-pr-comment-->

Dosant reviewed (approved):
Nice!

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/351503)
* Commit: dc069316c7f1b93f0c5579d8cca8fd2ac248f845

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/351503#0199e9c5-726e-47f7-9b98-cc91e2978049)
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/351503#0199e9c5-727a-4d57-ac44-ade9ea27d416)

### History
* :broken_heart: [Build #351455](https://buildkite.com/elastic/kibana-pull-request/builds/351455) failed a5bfe92df00802b948318c19e682393530717f9d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351503","number":351503,"commit":"dc069316c7f1b93f0c5579d8cca8fd2ac248f845"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351455","number":351455,"commit":"a5bfe92df00802b948318c19e682393530717f9d"}],"number":351503}
buildkite-pr-comment-->

jgowdyelastic reviewed (approved):
ML changes LGTM

TattdCodeMonkey committed:
chore: fix types in tests with addition of getInTrial

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/351900)
* Commit: e4b775e1bc29a160b5e37d41b0e82c27cc5b379e

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/351900#0199ee40-2116-4b77-a410-0bb0e4fc64f3)

### History
* :broken_heart: [Build #351503](https://buildkite.com/elastic/kibana-pull-request/builds/351503) failed dc069316c7f1b93f0c5579d8cca8fd2ac248f845
* :broken_heart: [Build #351455](https://buildkite.com/elastic/kibana-pull-request/builds/351455) failed a5bfe92df00802b948318c19e682393530717f9d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351900","number":351900,"commit":"e4b775e1bc29a160b5e37d41b0e82c27cc5b379e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351503","number":351503,"commit":"dc069316c7f1b93f0c5579d8cca8fd2ac248f845"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351455","number":351455,"commit":"a5bfe92df00802b948318c19e682393530717f9d"}],"number":351900}
buildkite-pr-comment-->

jbudz reviewed (approved):
`packages/kbn-optimizer/limits.yml` LGTM

jen-huang reviewed (approved):
Fleet changes LGTM

TattdCodeMonkey committed:
test: update to mock getInTrial

TattdCodeMonkey commented:
@elasticmachine merge upstream

jloleysens commented on src/platform/plugins/shared/navigation/public/plugin.test.ts:
```diff
@@ -193,11 +193,28 @@ describe('Navigation Plugin', () => {
     });
 
-    it('should set the feedback button visibility to "false" for deployment in trial', async () => {
+    it('should set the feedback button visibility to "false" for deployment in trial via endDate', async () => {
```
```suggestion
    it('should set the feedback button visibility to "false" for deployment in trial', async () => {
```

Seems like we're removing knowledge of "endDate"?

jloleysens commented on x-pack/platform/plugins/shared/cloud/public/plugin.test.ts:
```diff
@@ -199,6 +199,45 @@ describe('Cloud Plugin', () => {
+
+          expect(setup.getInTrial()).toBe(false);
+        });
+        it('is `false` when `serverless.in_trial` & `trial_end_date` are not set', () => {
```
What happens when both `serverless.in_trial` and `trial_end_date` are set? IMO we should throw and remove all `serverless.in_trial` test cases from this set of tests and scope those to the serverless:true config tests below.

jloleysens commented on x-pack/platform/plugins/shared/cloud/public/plugin.tsx:
```diff
@@ -120,6 +118,7 @@ export class CloudPlugin implements Plugin<CloudSetup, CloudStart> {
       ...this.cloudUrls.getUrls(), // TODO: Deprecate directly accessing URLs, use `getUrls` instead
       getPrivilegedUrls: this.cloudUrls.getPrivilegedUrls.bind(this.cloudUrls),
       getUrls: this.cloudUrls.getUrls.bind(this.cloudUrls),
+      getInTrial: this.isInTrial.bind(this),
```
Yeah, let's name it `isInTrial` to stick to the convention that methods returning booleans start with `is`.

jloleysens reviewed (approved):
Approving to unblock progress. Thanks for cleaning this up @TattdCodeMonkey.

I'd like to get your thoughts about removing `trialEndDate` from the setup contract before merging!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/354445)
* Commit: bc8f5b1036cfd07dbdbf13d8dcb16770769dcda7
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-239210)
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-239210-bc8f5b1036cf`
* [Security Deployment](https://kibana-pr-239210-security-b8c75e.kb.eu-west-1.aws.qa.elastic.cloud/login)
* [Elasticsearch Serverless Deployment](https://kibana-pr-239210-elasticsearch-e969a3.kb.eu-west-1.aws.qa.elastic.cloud/login)

### Failed CI Steps
* [FTR Configs #110](https://buildkite.com/elastic/kibana-pull-request/builds/354445#019a15d5-bdf9-489e-88cf-3af3845b7311)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354445#019a15d5-bdf9-489e-88cf-3af3845b7311) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354445/jobs/019a15d5-bdf9-489e-88cf-3af3845b7311/artifacts/019a15f3-7ecb-44aa-857b-6f07b71c7459) FTR Configs #<span></span>110 / ObservabilityApp Observability alerts pagination When 10 alerts are found Pagination controls Previous page button is disabled

### History
* :green_heart: [Build #353796](https://buildkite.com/elastic/kibana-pull-request/builds/353796) succeeded 30517acd130911b34867eb0ba5e2b02e3c6fac30
* :broken_heart: [Build #351942](https://buildkite.com/elastic/kibana-pull-request/builds/351942) failed 1189647f0e05d39ba1b50b02a1cfcc43649fc310
* :broken_heart: [Build #351900](https://buildkite.com/elastic/kibana-pull-request/builds/351900) failed e4b775e1bc29a160b5e37d41b0e82c27cc5b379e
* :broken_heart: [Build #351503](https://buildkite.com/elastic/kibana-pull-request/builds/351503) failed dc069316c7f1b93f0c5579d8cca8fd2ac248f845
* :broken_heart: [Build #351455](https://buildkite.com/elastic/kibana-pull-request/builds/351455) failed a5bfe92df00802b948318c19e682393530717f9d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/354445","number":354445,"commit":"bc8f5b1036cfd07dbdbf13d8dcb16770769dcda7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/353796","number":353796,"commit":"30517acd130911b34867eb0ba5e2b02e3c6fac30"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351942","number":351942,"commit":"1189647f0e05d39ba1b50b02a1cfcc43649fc310"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351900","number":351900,"commit":"e4b775e1bc29a160b5e37d41b0e82c27cc5b379e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351503","number":351503,"commit":"dc069316c7f1b93f0c5579d8cca8fd2ac248f845"}],"number":354445}
buildkite-pr-comment-->

TattdCodeMonkey committed:
cloud: rename getInTrial to isInTrial

TattdCodeMonkey committed:
test(cloud): improve grouping of isInTrial tests

TattdCodeMonkey committed:
test(cloud): fix broken tests

TattdCodeMonkey committed:
cloud: update limits and handle invalid trial end date

Worked on reducing page load bundle size for cloud, but still needed to
increase the limit slightly. While working on that change I also added
some extra code to handle if the configured `trial_end_date` does not
parse, which should never happen but is possible so I ultimately decided
to make sure it was handled and logged.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/355897)
* Commit: 14b5783adb29be787a3504d7875307d33712fb27

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/355897#019a2baf-c9a4-4458-851f-ca67dcebf5f6)

### History
* :broken_heart: [Build #355889](https://buildkite.com/elastic/kibana-pull-request/builds/355889) failed 54419eb902f923cce18c1fa95ae032538b6258ac
* :broken_heart: [Build #355476](https://buildkite.com/elastic/kibana-pull-request/builds/355476) failed 87075c8cf6054c0c1994a228e6556e81bffeee98
* :yellow_heart: [Build #354445](https://buildkite.com/elastic/kibana-pull-request/builds/354445) was flaky bc8f5b1036cfd07dbdbf13d8dcb16770769dcda7
* :green_heart: [Build #353796](https://buildkite.com/elastic/kibana-pull-request/builds/353796) succeeded 30517acd130911b34867eb0ba5e2b02e3c6fac30
* :broken_heart: [Build #351942](https://buildkite.com/elastic/kibana-pull-request/builds/351942) failed 1189647f0e05d39ba1b50b02a1cfcc43649fc310
* :broken_heart: [Build #351900](https://buildkite.com/elastic/kibana-pull-request/builds/351900) failed e4b775e1bc29a160b5e37d41b0e82c27cc5b379e

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/355897","number":355897,"commit":"14b5783adb29be787a3504d7875307d33712fb27"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/355889","number":355889,"commit":"54419eb902f923cce18c1fa95ae032538b6258ac"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/355476","number":355476,"commit":"87075c8cf6054c0c1994a228e6556e81bffeee98"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/354445","number":354445,"commit":"bc8f5b1036cfd07dbdbf13d8dcb16770769dcda7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/353796","number":353796,"commit":"30517acd130911b34867eb0ba5e2b02e3c6fac30"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351942","number":351942,"commit":"1189647f0e05d39ba1b50b02a1cfcc43649fc310"}],"number":355897}
buildkite-pr-comment-->
